### PR TITLE
Fixed #34619 -- Ensured FilteredSelectMultiple elements were associated with their labels.

### PR DESCRIPTION
**Pull request by GappleBee**
#### Trac ticket number

ticket-34619

#### Branch description
Added aria labels to elements in the admin:auth_group_add view

#### Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.

### CODE REVIEW

File: django/contrib/admin/static/admin/js/SelectFilter2.js [hunk 1]
```diff
@@ -15,6 +15,7 @@ Requires core.js and SelectBox.js.
             const from_box = document.getElementById(field_id);
             from_box.id += '_from'; // change its ID
             from_box.className = 'filtered';
+            from_box.setAttribute('aria-label', 'Available permissions');
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/SelectFilter2.js:18]: We need to make sure we translate any texts and this is also a general widget and not specific to permissions

**Review by sarahboyce — COMMENTED** [review]
Thank you for picking this up @GappleBee 

The approach needs tweaking here with particular attention to this [comment](https://code.djangoproject.com/ticket/34619#comment:3) from Thibaud which refers to this line:
```js
const title_available = quickElement('h2', selector_available, interpolate(gettext('Available %s') + ' ', [field_name]));
```

This [comment](https://code.djangoproject.com/ticket/34619#comment:9) is also quite important 

Note that this widget is not specific to permissions and be sure to make sure all texts are translated

**Comment by GappleBee** [comment]
@sarahboyce thanks for looking at my PR again, it seems you're the only one. I've revised the approach to pointing existing labels to their elements. Should I change the tests as well?

File: django/contrib/admin/static/admin/js/SelectFilter2.js [hunk 2]
```diff
@@ -69,32 +74,57 @@ Requires core.js and SelectBox.js.
...
-            choose_all.className = 'selector-chooseall';
+            const choose_all = quickElement(
+                'a',
+                selector_available,
+                interpolate(gettext('Choose all %s'), [field_name]),
+                'href', '#',
+                'role', 'button',
+                'id', field_id + '_add_all_link',
+                'class', 'selector-chooseall'
+            );
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/SelectFilter2.js:85]: There are a few "links" which are really buttons, if we can update the styling so that we can have these as proper buttons (so using the button tag and remove href # and role button) - that would be great
But that's more of a "bonus points" job and I imagine updating the css might be tricky

**Review by sarahboyce — COMMENTED** [review]
Thank you for the updates @GappleBee 
While I was testing with a screen reader, I did a few tweaks in this commit: ad241b192f23ec49a2ee5633c9562ec48e7b8ea2

The main idea is "title" is not accessible in general, not available on mobile, not announced to screen readers etc. If the information is important, it should be visible and so I have made the "help text" visible in the top of the boxes
image

Unfortunately, it made the boxes a little bigger and the plus add icon is no longer on the right - would you be able to fix the layout/css here?

You're right we should have a test 👍

File: django/contrib/admin/static/admin/js/SelectFilter2.js [hunk 3]
```diff
@@ -15,6 +15,7 @@ Requires core.js and SelectBox.js.
             const from_box = document.getElementById(field_id);
             from_box.id += '_from'; // change its ID
             from_box.className = 'filtered';
+            from_box.setAttribute('aria-labelledby', field_id + '_from_title');
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/SelectFilter2.js:18]: I have done this so the help text is also announced, as select boxes don't support `aria-describedby` very well 
`aria-describedby` didn't work on a select box for Narrator and I believe it might not work for JAWS either

**Comment by GappleBee** [comment]
Thanks for the changes @sarahboyce. I've changed the links to buttons and fixed the plus add icon. If there are no more issues,  should we start working on some tests?

File: django/contrib/admin/static/admin/css/widgets.css [hunk 4]
```diff
@@ -147,7 +156,7 @@
     background-position: 0 -80px;
 }
 
-a.selector-chooseall, a.selector-clearall {
+button.selector-chooseall, button.selector-clearall {
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/css/widgets.css:159]: We need to apply some of these changes in `django/contrib/admin/static/admin/css/rtl.css` 
So that the arrows change direction when we are using a RTL language

Comment (sarahboyce) [reply django/contrib/admin/static/admin/css/widgets.css:159]: Also thank you for updating the button :rocket:

File: django/contrib/admin/static/admin/js/SelectFilter2.js [hunk 5]
```diff
@@ -69,32 +74,55 @@ Requires core.js and SelectBox.js.
...
+            );
+            const remove_link = quickElement(
+                'a',
+                quickElement('li', selector_chooser),
+                interpolate(gettext('Remove selected chosen %s'), [field_name]),
+                'href', '#',
+                'id', field_id + '_remove_link',
+                'role', 'button',
+                'class', 'selector-remove'
+            );
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/SelectFilter2.js:105]:
```suggestion
            const add_link = quickElement(
                'button',
                quickElement('li', selector_chooser),
                interpolate(gettext('Choose selected %s'), [field_name]),
                'id', field_id + '_add_link',
                'class', 'selector-add'
            );
            const remove_link = quickElement(
                'button',
                quickElement('li', selector_chooser),
                interpolate(gettext('Remove selected chosen %s'), [field_name]),
                'id', field_id + '_remove_link',
                'class', 'selector-remove'
            );
```
Can you also look into the style (and potentially javascript) changes required to get this to work?

File: tests/admin_views/tests.py [hunk 6]
```diff
@@ -5801,6 +5801,34 @@ def setUp(self):
             title="A Long Title", published=True, slug="a-long-title"
         )
 
+    def test_add_group_aria_labels(self):
```
Comment (sarahboyce) [inline tests/admin_views/tests.py:5804]: In terms of what to test :thinking: maybe:
- the `aria-labelledby` id for the from and to box as well as the content of this (the label and help text)
- the `aria-label` for the filter boxes
- that the buttons are buttons not links

**Comment by sarahboyce** [comment]
@django/accessibility please also feel free to take a look and offer suggestions

File: tests/admin_views/tests.py [hunk 7]
```diff
@@ -5801,6 +5801,63 @@ def setUp(self):
...
+            "id_permissions_add_all_link",
+            "id_permissions_remove_all_link",
+            "id_permissions_add_link",
+            "id_permissions_remove_link"
+        )
+        
+        for link_id in link_ids:
+            link = self.selenium.find_element(By.ID, link_id)
+            self.assertEqual(link.tag_name, "button")
+
```
Comment (GappleBee) [inline tests/admin_views/tests.py:5860]: Just realised there was a js_tests file for testing SelectFilter2, I'll write the tests in that before removing these

**Comment by GappleBee** [comment]
Okay, I think the admin site looks nice now. @sarahboyce could you do a review of this?

File: django/contrib/admin/static/admin/js/SelectFilter2.js [hunk 8]
```diff
@@ -38,18 +39,22 @@ Requires core.js and SelectBox.js.
             // <div class="selector-available">
             const selector_available = quickElement('div', selector_div);
             selector_available.className = 'selector-available';
-            const title_available = quickElement('h2', selector_available, interpolate(gettext('Available %s') + ' ', [field_name]));
+            const selector_available_title = quickElement('div', selector_available);
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/SelectFilter2.js:42]: A selenium test needs fixing after this change, something like
```diff
--- a/tests/admin_inlines/tests.py
+++ b/tests/admin_inlines/tests.py
@@ -2420,31 +2420,41 @@ class SeleniumTests(AdminSeleniumTestCase):
             "admin:admin_inlines_courseproxy1_add",
             "admin:admin_inlines_courseproxy2_add",
         ]
-        css_selector = ".dynamic-class_set#class_set-%s h2"
+        css_available_selector = ".dynamic-class_set#class_set-%s .selector-available-title"
+        css_chosen_selector = ".dynamic-class_set#class_set-%s .selector-chosen-title"
 
         for url_name in tests:
             with self.subTest(url=url_name):
                 self.selenium.get(self.live_server_url + reverse(url_name))
                 # First inline shows the verbose_name.
-                available, chosen = self.selenium.find_elements(
-                    By.CSS_SELECTOR, css_selector % 0
+                available = self.selenium.find_element(
+                    By.CSS_SELECTOR, css_available_selector % 0
                 )
-                self.assertEqual(available.text, "AVAILABLE ATTENDANT")
-                self.assertEqual(chosen.text, "CHOSEN ATTENDANT")
+                chosen = self.selenium.find_element(
+                    By.CSS_SELECTOR, css_chosen_selector % 0
+                )
+                self.assertIn("Available attendant", available.text)
+                self.assertIn("Chosen attendant", chosen.text)
                 # Added inline should also have the correct verbose_name.
                 self.selenium.find_element(By.LINK_TEXT, "Add another Class").click()
-                available, chosen = self.selenium.find_elements(
-                    By.CSS_SELECTOR, css_selector % 1
+                available = self.selenium.find_element(
+                    By.CSS_SELECTOR, css_available_selector % 1
+                )
+                chosen = self.selenium.find_element(
+                    By.CSS_SELECTOR, css_chosen_selector % 1
                 )
-                self.assertEqual(available.text, "AVAILABLE ATTENDANT")
-                self.assertEqual(chosen.text, "CHOSEN ATTENDANT")
+                self.assertIn("Available attendant", available.text)
+                self.assertIn("Chosen attendant", chosen.text)
                 # Third inline should also have the correct verbose_name.
                 self.selenium.find_element(By.LINK_TEXT, "Add another Class").click()
-                available, chosen = self.selenium.find_elements(
-                    By.CSS_SELECTOR, css_selector % 2
+                available = self.selenium.find_element(
+                    By.CSS_SELECTOR, css_available_selector % 2
+                )
+                chosen = self.selenium.find_element(
+                    By.CSS_SELECTOR, css_chosen_selector % 2
                 )
-                self.assertEqual(available.text, "AVAILABLE ATTENDANT")
-                self.assertEqual(chosen.text, "CHOSEN ATTENDANT")
+                self.assertIn("Available attendant", available.text)
+                self.assertIn("Chosen attendant", chosen.text)
 
     def test_tabular_inline_layout(self):
```

Comment (GappleBee) [reply django/contrib/admin/static/admin/js/SelectFilter2.js:42]: Thanks, seems all the checks are good now 👍

File: django/contrib/admin/static/admin/css/rtl.css [hunk 9]
```diff
@@ -235,19 +235,19 @@ fieldset .fieldBox {
...
     background-position: 100% -144px;
 }
 
-a.selector-clearall {
+button.selector-clearall {
     background: url(../img/selector-icons.svg) 0 -160px no-repeat;
 }
 
-a.active.selector-clearall:focus, a.active.selector-clearall:hover {
+button.active.selector-clearall:focus, button.active.selector-clearall:hover {
```
Comment (knyghty) [inline django/contrib/admin/static/admin/css/rtl.css:250]: I kind of wonder if the `button` part is stopping something from breaking or if it's just extraneous. I know I've been trying to avoid using elements as part of the selector where possible.

**Review by knyghty — COMMENTED** [review]
I left a small comment but I'm mostly here to say I'll try to test this during the week or on the weekend. It looks like a clear improvement but I'd like to try it in action.

**Comment by knyghty** [comment]
I did spot one thing that doesn't seem quite right to me. It seems there are two help texts for this type of field. One that tells you how to use the widget, and one how to choose multiple options. Ideally, both of these should be (if screen readers support this - I didn't really check). Per the spec we can set this to a list of IDs: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby#values

So it would probably be good to give both these help texts a unique ID and associate them both.

I'm also unsure if we should use the word "clicking" as this doesn't really apply for screen reader users that are probably also keyboard users. Perhaps they know what this means though, as it's used quite widely. I don't think there is really a need to change this now, just to keep in mind.

**Comment by sarahboyce** [comment]
> I did spot one thing that doesn't seem quite right to me. It seems there are two help texts for this type of field. One that tells you how to use the widget, and one how to choose multiple options. Ideally, both of these should be (if screen readers support this - I didn't really check). Per the spec we can set this to a list of IDs: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby#values

I think this is a symptom of a different problem which is admin forms ignores the widget `use_fieldset` attribute
This doesn't just impact `RelatedFieldWidgetWrapper` but also all other widgets which should use a fieldset and a legend like `SelectDateWidget`

As a very rough proof of concept, we can probably add in support for this:
```diff
diff --git a/django/contrib/admin/helpers.py b/django/contrib/admin/helpers.py
index d28a382814..3d8a31db30 100644
--- a/django/contrib/admin/helpers.py
+++ b/django/contrib/admin/helpers.py
@@ -173,6 +173,7 @@ class AdminField:
         self.is_first = is_first  # Whether this field is first on the line
         self.is_checkbox = isinstance(self.field.field.widget, forms.CheckboxInput)
         self.is_readonly = False
+        self.is_fieldset = self.field.field.widget.use_fieldset
 
     def label_tag(self):
         classes = []
@@ -185,12 +186,14 @@ class AdminField:
         if not self.is_first:
             classes.append("inline")
         attrs = {"class": " ".join(classes)} if classes else {}
+        tag = "legend" if self.is_fieldset else None
         # checkboxes should not have a label suffix as the checkbox appears
         # to the left of the label.
         return self.field.label_tag(
             contents=mark_safe(contents),
             attrs=attrs,
             label_suffix="" if self.is_checkbox else None,
+            tag=tag,
         )
 
     def errors(self):
diff --git a/django/contrib/admin/templates/admin/includes/fieldset.html b/django/contrib/admin/templates/admin/includes/fieldset.html
index 9c9b31965a..9cd8cc6a68 100644
--- a/django/contrib/admin/templates/admin/includes/fieldset.html
+++ b/django/contrib/admin/templates/admin/includes/fieldset.html
@@ -12,6 +12,7 @@
             {% if line.fields|length == 1 %}{{ line.errors }}{% else %}<div class="flex-container form-multiline">{% endif %}
             {% for field in line %}
                 <div>
+                    {% if field.is_fieldset %}<fieldset{% if field.field.help_text %} aria-describedby="{{ field.field.id_for_label }}_helptext"{% endif %}>{% endif %}
                     {% if not line.fields|length == 1 and not field.is_readonly %}{{ field.errors }}{% endif %}
                         <div class="flex-container{% if not line.fields|length == 1 %} fieldBox{% if field.field.name %} field-{{ field.field.name }}{% endif %}{% if not field.is_readonly and field.errors %} errors{% endif %}{% if field.field.is_hidden %} hidden{% endif %}{% endif %}{% if field.is_checkbox %} checkbox-row{% endif %}">
                             {% if field.is_checkbox %}
@@ -30,6 +31,7 @@
                             <div>{{ field.field.help_text|safe }}</div>
                         </div>
                     {% endif %}
+                    {% if field.is_fieldset %}</fieldset>{% endif %}
                 </div>
             {% endfor %}
             {% if not line.fields|length == 1 %}</div>{% endif %}
diff --git a/django/contrib/admin/widgets.py b/django/contrib/admin/widgets.py
index 74004a7ba7..0e49266c06 100644
--- a/django/contrib/admin/widgets.py
+++ b/django/contrib/admin/widgets.py
@@ -250,6 +250,7 @@ class RelatedFieldWidgetWrapper(forms.Widget):
     """
 
     template_name = "admin/widgets/related_widget_wrapper.html"
+    use_fieldset = True
 
     def __init__(
```
I say "rough" as the styling needs adjusting, tests etc
screenshot of some resulting widget fieldsets which are styled wrong

In short, this widget has multiple inputs and so technically should be a fieldset with an appropriate `aria-describedby` attribute for the field help text and the label should be a legend.

> Though to be honest I am not sure if it falls in the scope of this PR.

I think it deserves it's own ticket as it's not specific to this widget

**Comment by knyghty** [comment]
@sarahboyce Thanks, I agree. Everything else looks fine as far as I could tell, so I'm happy 👍🏻

**Comment by sarahboyce** [comment]
> I'm also unsure if we should use the word "clicking" as this doesn't really apply for screen reader users that are probably also keyboard users. Perhaps they know what this means though, as it's used quite widely. I don't think there is really a need to change this now, just to keep in mind.

I am not strongly attached to the helptext, we can update these :+1:

**Review by sarahboyce — APPROVED** [review]
Thank you @GappleBee ⭐ 
I pushed minor tweaks and did another test with a screen reader - this looks good to me 👍
